src/doc-utils.js

```diff
@@ -1,34 +1,35 @@
 "use strict";
 
 function traverseDoc(doc, onEnter, onExit, shouldTraverseConditionalGroups) {
-  var hasStopped = false;
   function traverseDocRec(doc) {
+    var shouldRecurse = true;
     if (onEnter) {
-      hasStopped = hasStopped || onEnter(doc) === false;
-    }
-    if (hasStopped) {
-      return;
+      if (onEnter(doc) === false) {
+        shouldRecurse = false;
+      }
     }
 
-    if (doc.type === "concat") {
-      for (var i = 0; i < doc.parts.length; i++) {
-        traverseDocRec(doc.parts[i]);
-      }
-    } else if (doc.type === "if-break") {
-      if (doc.breakContents) {
-        traverseDocRec(doc.breakContents);
-      }
-      if (doc.flatContents) {
-        traverseDocRec(doc.flatContents);
-      }
-    } else if (doc.type === "group" && doc.expandedStates) {
-      if (shouldTraverseConditionalGroups) {
-        doc.expandedStates.forEach(traverseDocRec);
-      } else {
+    if (shouldRecurse) {
+      if (doc.type === "concat") {
```

Comment:
https://github.com/prettier/prettier/pull/1294/files?w=1 makes github ignore whitespace in diffs. Sadly you can't comment, but at least it makes it readable 😄